lib/sms/sms.c

```diff
@@ -0,0 +1,322 @@
+/*
+ * Copyright (c) 2019 Nordic Semiconductor ASA
+ *
+ * SPDX-License-Identifier: LicenseRef-BSD-5-Clause-Nordic
+ */
+
+#include <logging/log.h>
+#include <zephyr.h>
+#include <stdio.h>
+#include <sms.h>
+#include <errno.h>
+#include <at_cmd.h>
+#include <at_cmd_parser/at_cmd_parser.h>
+#include <at_cmd_parser/at_params.h>
+#include <at_notif.h>
+
+LOG_MODULE_REGISTER(sms, CONFIG_SMS_LOG_LEVEL);
+
+#define AT_SMS_PARAMS_COUNT_MAX 6
+#define AT_SMS_RESPONSE_MAX_LEN 256
+
+#define AT_CNMI_PARAMS_COUNT 6
+#define AT_CMT_PARAMS_COUNT 4
+
+/** @brief AT command to check if a client already exist. */
+#define AT_SMS_SUBSCRIBER_READ "AT+CNMI?"
+
+/** @brief AT command to register an SMS client. */
+#define AT_SMS_SUBSCRIBER_REGISTER "AT+CNMI=3,2,0,1"
+
+/** @brief AT command to unregister an SMS client. */
+#define AT_SMS_SUBSCRIBER_UNREGISTER "AT+CNMI=0,0,0,0"
+
+/** @brief AT command to an ACK in PDU mode. */
+#define AT_SMS_PDU_ACK "AT+CNMA=1"
+
+/** @brief Start of AT notification for incoming SMS. */
+#define AT_SMS_NOTIFICATION "+CMT:"
+#define AT_SMS_NOTIFICATION_LEN (sizeof(AT_SMS_NOTIFICATION) - 1)
+
+static struct at_param_list resp_list;
+static char resp[AT_SMS_RESPONSE_MAX_LEN];
+
+/**
+ * @brief Indicates that the module has been successfully initialized
+ * and registered as an SMS client.
+ */
+static bool sms_client_registered;
+
+/** @brief SMS PDU data. */
+struct sms_data {
+	char *alpha;
+	u16_t length;
+	char *pdu;
+};
+
+/** @brief SMS event. */
+static struct sms_data cmt_rsp;
+
+struct sms_subscriber {
+	/* Listener user context. */
+	void *ctx;
+	/* Listener callback. */
+	sms_callback_t listener;
+};
+
+/** @brief List of subscribers. */
+static struct sms_subscriber subsribers[CONFIG_SMS_MAX_SUBSCRIBERS_CNT];
+
+/** @brief Check if the received AT notification is a +CMT event. */
+static bool sms_is_cmd_notification(const char *const at_notif)
+{
+	if (at_notif == NULL) {
+		return false;
+	}
+
+	if (strlen(at_notif) <= AT_SMS_NOTIFICATION_LEN) {
+		return false;
+	}
+
+	return (strncmp(at_notif, AT_SMS_NOTIFICATION,
+			AT_SMS_NOTIFICATION_LEN) == 0);
+}
+
+/** @brief Parse the +CMT unsolicited received message in PDU mode. */
+static int sms_cmt_notif_parse(const char *const buf)
+{
+	/* Parse the received message. */
+	int err = at_parser_max_params_from_str(buf, NULL, &resp_list,
+						AT_CMT_PARAMS_COUNT);
+
+	if (err != 0) {
+		LOG_ERR("Unable to parse CMT notification, err=%d", err);
+		return err;
+	}
+
+	/* Check the AT response format. */
+	if (at_params_valid_count_get(&resp_list) != AT_CMT_PARAMS_COUNT) {
+		LOG_ERR("Invalid CMT notification format");
+		return -EAGAIN;
+	}
+
+	return 0;
+}
+
+/** @brief Save the SMS notification parameters. */
+static int sms_cmt_notif_save(void)
+{
+	if (cmt_rsp.alpha != NULL) {
+		k_free(cmt_rsp.alpha);
+	}
+
+	if (cmt_rsp.pdu != NULL) {
+		k_free(cmt_rsp.pdu);
+	}
+
+	/* Save alpha as a null-terminated String. */
+	size_t alpha_len;
```

Comment:
I would put declarations at the beginning of a function, it's easier to see how much data is allocated on the stack and to get an idea of what the function will try to do next.